zkml/src/layers/transformer/layernorm.rs

```diff
@@ -1,33 +1,260 @@
+        // Finally we have to multiply all of this by `multiplier = LAYERNORM_SCALE_FACTOR * input_scaling.scale() * input_scaling.scale()` so we have `ceil_log2(multiplier)`
+        // additional bits on top of this.
+
+        // Get the input scale
+        let input_scale = input_scaling.scale();
+        // Get the dim size (N)
+        let dim_size = self.normalisation_dim_size();
+        // We work out what we have to mutliply by so that everything is scaled to `LAYERNORM_SCALE_FACTOR` in quantised world
+        let multiplier =
+            (LAYERNORM_SCALE_FACTOR as f32 * input_scale * input_scale).round() as Element;
+        // Work out the number of variables the table requires, this is likely to be far too large to actually materialise as a table
+        let full_table_bit_size = 2 * (ceil_log2(dim_size) + *quantization::BIT_LEN - 1)
+            + ceil_log2(multiplier as usize)
+            + 1;
+        // To get around this we use the fact that we should only have roughly `2*(*quantization::BIT_LEN -1)` bits of precision i.e. only the most significant `2*(*quantization::BIT_LEN -1)`
+        // can actually be "trusted" the rest are essentially junk because they don't come from the actual inputs and are just guesses at the part that we have alread "rounded away" in quantisation.
+        // So the actual part we perform inverse square root on is size `2*(*quantization::BIT_LEN -1)` and then we rust need the discarded part to be range checked.
+        let range_checked_bits = full_table_bit_size - 2 * (*quantization::BIT_LEN - 1);
+
+        // The final chunk might be values with fewer than *quantization::BIT_LEN bits so we work out what we need to scale the value up by in order to use our standard range check table.
+        let remainder_bits = range_checked_bits % *quantization::BIT_LEN;
+        let top_chunk_scalar_log = if !remainder_bits.is_zero() {
+            *quantization::BIT_LEN - remainder_bits
+        } else {
+            0
+        };
+        // Calculate the lookup table
+        let table_max: Element = 1 << (2 * (*quantization::BIT_LEN - 1));
+        let table_min = -table_max;
+        // Because we don't use the same formula for the standard deviation as LayerNorm does in float we have to rescale `self.eps` in this case to be `N^2 * self.eps`
+        let rescaled_eps = (dim_size * dim_size) as f32 * self.eps;
+        let lut = (table_min..table_max)
+            .map(|val| {
+                // First we have to shift by `range_checked_bits`
+                let shifted_val = val << range_checked_bits;
+                // Now we convert back to float and perform the operation
+                let float_output = 1.0f32
+                    / ((shifted_val as f32 / LAYERNORM_SCALE_FACTOR as f32) + rescaled_eps).sqrt();
+                // Now we use the output scale factor to recover the element value
+                (float_output * LAYERNORM_OUTPUT_SCALE_FACTOR as f32).round() as Element
```

Comment:
Same here - can you extract a single function that does the table and then use it everywhere ?